packages/itwinui-react/src/core/Panels/Panels.tsx

```diff
@@ -0,0 +1,600 @@
+    animations == null
+      ? id !== currentPanelId?.id
+      : !Object.keys(animations).includes(id);
+
+  const isInert = isHidden || isAnimating;
+
+  const associatedTrigger = !!id ? triggers?.current?.get(id) : undefined;
+
+  return (
+    <PanelContext.Provider
+      value={{
+        id,
+        associatedTrigger,
+      }}
+    >
+      <Box
+        ref={useMergedRefs(setElement, forwardedRef)}
+        id={id}
+        aria-labelledby={`${id}-header`}
+        tabIndex={-1}
+        {...rest}
+        className={cx('iui-panel', className, {
+          'iui-panel-visible': id === currentPanelId?.id,
+        })}
+        style={{
+          display: isHidden ? 'none' : undefined,
+
+          // Add the last keyframe styles to the current panel to avoid flickering
+          // i.e. showing the current panel for a split second between when the animations ends and the panel is hidden
+          ...(id === currentPanelId?.id &&
+          animations != null &&
+          animations[id] != null
+            ? animations[id][animations[id].length - 1]
+            : {}),
+
+          ...style,
+        }}
+        {...{ inert: isInert ? '' : undefined }}
+      >
+        {children}
+      </Box>
+    </PanelContext.Provider>
+  );
+}) as PolymorphicForwardRefComponent<'div', PanelProps>;
+Panel.displayName = 'Panels.Panel';
+
+const PanelContext = React.createContext<
+  | {
+      id: string;
+      associatedTrigger: TriggerMapEntry | undefined;
+    }
+  | undefined
+>(undefined);
+
+// ----------------------------------------------------------------------------
+
+type PanelTriggerProps = {
+  for: string;
+  children: React.ReactElement;
+};
+
+/**
+ * Wraps the button to append an `onClick` that changes the `activeId` to the `for` prop.
+ * This component also creates an internal triggers map that stores a link between a <trigger + its current panel> and
+ * the panel it points to. This is useful for back navigation (see `Panels.Header`).
+ *
+ * @example
+ * <Panels.Panel id={panelIdRoot} as={List}>
+ *   <Surface.Header as={Panels.Header}>Root</Surface.Header>
+ *   <ListItem>
+ *     <Panels.Trigger for={panelIdMoreInfo}>
+ *       <ListItem.Action>More details</ListItem.Action>
+ *     </Panels.Trigger>
+ *   </ListItem>
+ * </Panels.Panel>
+ */
+const PanelTrigger = (props: PanelTriggerProps) => {
+  const { children, for: forProp } = props;
+
+  const panelWrapperContext = React.useContext(PanelsWrapperContext);
+  const panelContext = React.useContext(PanelContext);
+
+  const fallbackId = React.useId();
+  const triggerId = children.props.id || fallbackId;
+
+  const onClick = React.useCallback(() => {
+    panelWrapperContext?.changeActivePanel({
+      newActiveId: {
+        id: forProp,
+        direction: 'next',
+      },
+    });
+  }, [forProp, panelWrapperContext]);
+
+  if (!React.isValidElement(children)) {
+    return null;
+  }
+
+  const triggersMatch = panelWrapperContext?.triggers.current?.get(forProp);
```

Comment:
Reading refs during rendering is an anti-pattern. Why can't this be a state?

Reply:
Based on your link, floating-ui is _not_ reading/writing refs during rendering. It is ok to read/write refs (via `.current`) inside Effects, callback refs and event handlers. It is *not* ok to read/write refs (via `.current`) directly in the rendering path. See [docs](https://react.dev/learn/referencing-values-with-refs#best-practices-for-refs).